src/main/java/com/teragrep/rsm_01/ErrorCallbackImpl.java

```
@@ -0,0 +1,81 @@
+ * section 7
+ *
+ * If you modify this Program, or any covered work, by linking or combining it
+ * with other code, such other code is not for that reason alone subject to any
+ * of the requirements of the GNU Affero GPL version 3 as long as this Program
+ * is the same Program as licensed from Suomen Kanuuna Oy without any additional
+ * modifications.
+ *
+ * Supplemented terms under GNU Affero General Public License version 3
+ * section 7
+ *
+ * Origin of the software must be attributed to Suomen Kanuuna Oy. Any modified
+ * versions must be marked as "Modified version of" The Program.
+ *
+ * Names of the licensors and authors may not be used for publicity purposes.
+ *
+ * No rights are granted for use of trade names, trademarks, or service marks
+ * which are in The Program if any.
+ *
+ * Licensee must indemnify licensors and authors for any liability that these
+ * contractual assumptions impose on licensors and authors.
+ *
+ * To the extent this program is licensed as part of the Commercial versions of
+ * Teragrep, the applicable Commercial License may apply to this file if you as
+ * a licensee so wish it.
+ */
+package com.teragrep.rsm_01;
+
+import com.sun.jna.Pointer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+
+public final class ErrorCallbackImpl implements ErrorCallback {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ErrorCallbackImpl.class);
+
+    private final ArrayList<String> errors;
+
+    public ErrorCallbackImpl() {
+        errors = new ArrayList<>();
+    }
+
+    @Override
+    public void invoke(Pointer cookie, String msg, int length) {
+        errors.add(msg);
+        LOGGER.error("liblognorm error: {}", msg);
+    }
+
+    /**
+     * Method for throwing any received liblognorm error messages as java exception.
+     */
+    public void throwOccurredErrors() {
+        if (!errors.isEmpty()) {
```

should thrown errors be cleared or are they always so fatal that the object instance is invalid after some occured?

---

The rulebase loading method for liblognorm will not return error codes when the rulebase has errors in it, instead it will generate error messages and return 0 indicating that the loading of the rulebase was successful.
By implementing the check for error messages those rulebase errors can be caught as java exceptions.

In my opinion the errors should be handled using fail fast, making the object instance invalid if they occur.

---

Here is an example of how the errors can completely break the rulebase while no error codes are returned by liblognorm during normalization:

Partially invalid rulebase with a typo on last line:
`version=2
rule=:%all:rest%
ruletag:Amount: %N:number%`

Message to normalize:
`Amount: 1234`

Normalization result with the invalid rulebase is:
`{ "all": "Amount: 1234" }`

But with a valid rulebase:
`version=2
rule=:%all:rest%
#Amount: 1234
rule=tag:Amount: %N:number%`

The normalization result is:
`{ "N": "1234", "event.tags": [ "tag" ] }`